File: src/v8/experts/bollinger_reversion.py
"""Bollinger band-reversion behavior family (`bollinger_reversion`).

Hypothesis (mechanism `volatility_band`, behavior `band_reversion`): price
testing the 2-SD band tends to revert toward the 1-SD band / central value —
Setup 2 fades the 2-SD band back to the 1-SD band with the stop beyond the
3-SD band (Ch12 p481-482); Setup 3 fades the SAME band structure but only in
the direction of the established trend (close above the SMA in an UPTREND ->
long at the upper 2-SD band, target the 1-SD band; Ch12 p482). The opposite
hypothesis — riding the band — is the sibling behavior `band_breakout`
(`bollinger_breakout`) under the SAME `volatility_band` mechanism (rule 13).

Book geometry (verbatim): Setup 2 — short at the upper 2-SD band, stop just
above the 3-SD upper band, profit target at the 1-SD upper band (mirror for
the long leg; the book text's long leg reads "upper two-SD band" but the
"just below the three-SD lower band" stop implies the lower 2-SD band — the
typo is captured verbatim in the extraction). Setup 3 — stop just under/over
the SMA, profit exit at the 1-SD band. The three levels are one band-sigma
apart, so in R terms stop_r = target_r = sigma_ref / atr_ref for Setup 2 and
stop_r = 2 * sigma_ref / atr_ref, target_r = sigma_ref / atr_ref for Setup 3.
A close beyond the 3-SD band invalidates the reversion premise (that is a
trend, not a reversion; Ch12 p471-474).

Setup 3 ships RR = 0.5 (stop_r = 2*sd/atr, target_r = sd/atr): at
round_trip_cost_r = 0.07 the breakeven win rate is 2.07/3.00 = 69.0% — a
high-hit-rate/low-RR geometry, recorded as a PROVISIONAL_DECISION (issue #70;
D-061). Setup 2 and Setup 3 outcomes are aggregated in the outcome ledger
(variant not recorded per outcome today), so Setup 3's standalone
hit-rate-vs-breakeven must be measured before this geometry is relied on.

As in `bollinger_breakout`, the band geometry is FROZEN at the setup anchor
(run start, D-026) so episode_key stays stable across the detection run; the
20 closes ending at a fixed anchor are identical on every decision clock while
the anchor stays inside the 32-bar history window.
"""
from __future__ import annotations

from ..schema import MarketState, CandidateDraft, ExpertEvaluation
from .base import Expert

# Declared, frozen constant (D-036 pattern): the Bollinger base window of the
# marketstate bb_* features (SMA20 +/- 2*sigma); the 1-SD/3-SD levels are
# derived from it, not new features.
BB_BASE_N = 20


def _mean(values):
    return sum(values) / len(values)


def _std_pop(values):
    m = _mean(values)
    return (sum((v - m) ** 2 for v in values) / len(values)) ** 0.5


def _bb_series(hist):
    """Per-bar (mid, sd) of the trailing 20 closes; None in warmup."""
    closes = [b[4] for b in hist]
    out = []
    for i in range(len(closes)):
        if i >= BB_BASE_N - 1:
            win = closes[i - BB_BASE_N + 1:i + 1]
            out.append((_mean(win), _std_pop(win)))
        else:
            out.append(None)
    return out


def _anchor_refs(hist, anchor_event_id):
    """Frozen band stack + ATR14 at the setup anchor, or None when the anchor's
    20-bar context is not fully inside the history window (the documented
    anchor bound of base.Expert.find_setup_anchor)."""
    pos = next((i for i, b in enumerate(hist) if b[0] == anchor_event_id), None)
    if pos is None or pos < BB_BASE_N - 1 or pos < 13:
        return None
    closes = [b[4] for b in hist]
    win = closes[pos - BB_BASE_N + 1:pos + 1]
    mid = _mean(win)
    sd = _std_pop(win)
    atr = _mean([hist[k][2] - hist[k][3] for k in range(pos - 13, pos + 1)])
    return {'mid_ref': mid, 'sd_ref': sd, 'atr_ref': atr,
            'upper_1sd_ref': mid + sd, 'upper_2sd_ref': mid + 2 * sd,
            'upper_3sd_ref': mid + 3 * sd,
            'lower_1sd_ref': mid - sd, 'lower_2sd_ref': mid - 2 * sd,
            'lower_3sd_ref': mid - 3 * sd}


class BollingerReversionExpert(Expert):
    """Fade the 2-SD band back toward the 1-SD band — Setup 2 (reversion
    within bands) or Setup 3 (trend-aligned reversion)."""
    expert_id = 'bollinger_reversion'
    version = 'v1'
    mechanism_family_id = 'volatility_band'
    behavior_family_id = 'band_reversion'
    variant_id = 'a'
    requires = ('trend', 'volatility', 'history')
    # D-044: the full evaluated set, losers included. `a` = Setup 2 (fade the
    # 2-SD band, stop beyond 3-SD, target 1-SD); `b` = Setup 3 (trend-aligned
    # reversion: close beyond the SMA in the trend direction).
    variants_evaluated = ('a', 'b')
    search_universe_size = 2

    def __init__(self, variant_id=None):
        if variant_id is not None:
            if variant_id not in self.variants_evaluated:
                raise ValueError(
                    f'{self.expert_id}: unknown variant {variant_id!r}; '
                    f'evaluated variants: {self.variants_evaluated}')
            self.variant_id = variant_id

    # --- per-variant per-bar predicates (D-026 anchor scan) ------------------
    def _pred_a_long(self, i, bar):
        """Close in the fade zone below the lower band: between 2-SD and 3-SD
        below the SMA. A degenerate band (sd == 0) is no fade level (with
        upper == lower == mid the inequality holds for every bar), and a close
        beyond 3-SD is a breakdown, not a reversion (Ch12 p471-474)."""
        if i < BB_BASE_N - 1:
            return False
        mid, sd = self._bb[i]
        return sd > 0 and mid - 3 * sd < bar[4] <= mid - 2 * sd

    def _pred_a_short(self, i, bar):
        """Close in the fade zone above the upper band: between 2-SD and 3-SD
        above the SMA (a close beyond 3-SD is a breakout, not a reversion)."""
        if i < BB_BASE_N - 1:
            return False
        mid, sd = self._bb[i]
        return sd > 0 and mid + 2 * sd <= bar[4] < mid + 3 * sd

    def _pred_b_long(self, i, bar):
        """Close above the SMA in an UPTREND (Setup 3, Ch12 p482)."""
        if i < BB_BASE_N - 1:
            return False
        _e, _o, _h, _l, close, ema_fast, ema_slow = bar
        return close > self._bb[i][0] and ema_fast > ema_slow

    def _pred_b_short(self, i, bar):
        """Close below the SMA in a DOWNTREND."""
        if i < BB_BASE_N - 1:
            return False
        _e, _o, _h, _l, close, ema_fast, ema_slow = bar
        return close < self._bb[i][0] and ema_fast < ema_slow

    def _pred(self, direction):
        if self.variant_id == 'a':
            return (self._pred_a_long if direction == 'LONG'
                    else self._pred_a_short)
        return (self._pred_b_long if direction == 'LONG'
                else self._pred_b_short)

    def _direction(self, f, sym, close):
        if self.variant_id == 'a':
            # Fade zone: close between 2-SD and 3-SD (matches _pred_a_*).
            mid = float(f[f'{sym}.bb_mid'].value)
            upper = float(f[f'{sym}.bb_upper'].value)
            lower = float(f[f'{sym}.bb_lower'].value)
            upper_3sd = mid + 1.5 * (upper - mid)
            lower_3sd = mid - 1.5 * (mid - lower)
            if upper <= close < upper_3sd:
                return 'SHORT', 'upper_2sd_ref'
            if lower_3sd < close <= lower:
                return 'LONG', 'lower_2sd_ref'
            return None, ''
        fast = float(f[f'{sym}.ema_fast'].value)
        slow = float(f[f'{sym}.ema_slow'].value)
        mid = float(f[f'{sym}.bb_mid'].value)
        if close > mid and fast > slow:
            return 'LONG', 'mid_ref'
        if close < mid and fast < slow:
            return 'SHORT', 'mid_ref'
        return None, ''

    def _geometry(self, refs, direction):
        sd, atr = refs['sd_ref'], refs['atr_ref']
        geo = {'entry': 'NEXT_BAR_CLOSE', 'expiry_bars': 8, 'atr_ref': atr,
               'variant': self.variant_id}
        if self.variant_id == 'b':
            # Setup 3 (Ch12 p482): entry proxy at the 2-SD band, stop under
            # the SMA (two sigma), profit exit at the 1-SD band (one sigma).
            geo['stop_r'] = 2 * sd / atr
            geo['target_r'] = sd / atr
            geo['mid_ref'] = refs['mid_ref']
            if direction == 'LONG':
                geo['upper_1sd_ref'] = refs['upper_1sd_ref']
                geo['upper_2sd_ref'] = refs['upper_2sd_ref']
            else:
                geo['lower_1sd_ref'] = refs['lower_1sd_ref']
                geo['lower_2sd_ref'] = refs['lower_2sd_ref']
            return geo
        # Setup 2 (Ch12 p481-482): fade the 2-SD band; the 3-SD stop and the
        # 1-SD target are each one band-sigma away.
        r = sd / atr
        geo['stop_r'] = r
        geo['target_r'] = r
        if direction == 'SHORT':
            geo['upper_1sd_ref'] = refs['upper_1sd_ref']
            geo['upper_2sd_ref'] = refs['upper_2sd_ref']
            geo['upper_3sd_ref'] = refs['upper_3sd_ref']
        else:
            geo['lower_1sd_ref'] = refs['lower_1sd_ref']
            geo['lower_2sd_ref'] = refs['lower_2sd_ref']
            geo['lower_3sd_ref'] = refs['lower_3sd_ref']
        return geo

    def evaluate(self, state: MarketState) -> ExpertEvaluation:
        t = state.as_of
        sym = state.universe[0]
        need = [f'{sym}.close', f'{sym}.bb_mid', f'{sym}.bb_upper',
                f'{sym}.bb_lower', f'{sym}.bb_pct_b', f'{sym}.ema_fast',
                f'{sym}.ema_slow', f'{sym}.history']
        if not self._need(state, need):
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        f = state.features
        close = float(f[f'{sym}.close'].value)
        hist_value = f[f'{sym}.history'].value
        if not isinstance(hist_value, (tuple, list)) or not hist_value:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        self._hist = tuple(hist_value)
        self._bb = _bb_series(self._hist)
        direction, ref_key = self._direction(f, sym, close)
        if direction is None:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_SETUP', t)
        anchor = self.find_setup_anchor(self._hist, self._pred(direction))
        refs = _anchor_refs(self._hist, anchor)
        if refs is None or refs['sd_ref'] <= 0 or refs['atr_ref'] <= 0:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        geometry = self._geometry(refs, direction)
        draft = CandidateDraft(
            expert_id=self.expert_id, expert_version=self.version,
            instrument=sym, direction=direction,
            setup_fingerprint=f'{sym}:{close:.6f}:{geometry[ref_key]:.6f}',
            risk_geometry=geometry,
            birth_time=t, setup_anchor_event_id=anchor)
        return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                'APPLICABLE', 'CANDIDATE', t, draft)

    def still_valid(self, state: MarketState, draft: CandidateDraft) -> bool:
        """Post-entry thesis: the reversion premise. Setup 2 dies when price
        closes beyond the frozen 3-SD band (that is a trend, not a reversion);
        Setup 3 dies when the trend flips. Unobservable inputs fail open.
        """
        sym = draft.instrument
        f = state.features
        close = f.get(f'{sym}.close')
        if close is None or close.value is None:
            return True
        c = float(close.value)
        long = draft.direction == 'LONG'
        geom = draft.risk_geometry
        if geom.get('variant') == 'b':
            fast = f.get(f'{sym}.ema_fast')
            slow = f.get(f'{sym}.ema_slow')
            if fast is None or slow is None or fast.value is None \
                    or slow.value is None:
                return True
            return float(fast.value) > float(slow.value) if long \
                else float(fast.value) < float(slow.value)
        ref = geom.get('upper_3sd_ref' if not long else 'lower_3sd_ref')
        if ref is None:
            return True
        return c < float(ref) if not long else c > float(ref)

File: v8-core/src/experts/bollinger_reversion.rs
//! bollinger_reversion: fade the 2-SD band back toward the 1-SD band (Setup 2
//! / Setup 3) — mirror src/v8/experts/bollinger_reversion.py bit-for-bit
//! (PARITY_AND_IDENTITY_SPEC §3; COMPUTE_CORE_SPEC §8 S4). Default instance is
//! variant 'a' (the Python oracle's default constructor).

use crate::experts::base::*;
use crate::simulator::Draft;
use crate::state::{fsum, HistBar};

pub const PORTED: bool = true;
pub const VERSION: &str = "v1";
pub const REQUIRES: &[&str] = &["trend", "volatility", "history"];
// Declared risk geometry (EXPERT_PROTOCOL §1: risk geometry is "Predeclared
// entry, stop, target, timeout and sizing inputs"; SIMULATION_TRUTH_SPEC D-028:
// R is a declared price distance). Fixed values are declared here, never
// re-literalized inside evaluate(); a structural target/stop is computed at
// the call site and overrides the key.
pub const EXPIRY_BARS: i64 = 8;

/// Declared, frozen constant (D-036 pattern): the Bollinger base window of the
/// marketstate bb_* features (SMA20 +/- 2*sigma); the 1-SD/3-SD levels are
/// derived from it, not new features.
const BB_BASE_N: usize = 20;

/// Python `_mean`: `sum(values) / len(values)` — CPython's sum() over floats
/// is compensated (state::fsum is bit-identical, per state.rs).
fn mean(values: &[f64]) -> f64 {
    fsum(values) / values.len() as f64
}

/// Python `_std_pop`: `(sum((v - m) ** 2 for v in values) / len(values)) ** 0.5`
/// — both powers are libm pow (x**2 / x**0.5), not x*x / sqrt. black_box keeps
/// the exponents opaque so LLVM cannot fold pow -> x*x / sqrt in release
/// (same pattern as state.rs std_pop).
fn std_pop(values: &[f64]) -> f64 {
    let m = mean(values);
    let mut acc = Vec::with_capacity(values.len());
    for v in values {
        let d = v - m;
        acc.push(d.powf(std::hint::black_box(2.0)));
    }
    let sos = fsum(&acc);
    (sos / values.len() as f64).powf(std::hint::black_box(0.5))
}

/// Python `_bb_series`: per-bar (mid, sd) of the trailing 20 closes; None in
/// warmup (entry i<BB_BASE_N-1 is a placeholder, never read — every predicate
/// guards i >= BB_BASE_N-1 first).
fn bb_series(hist: &[HistBar]) -> Vec<(f64, f64)> {
    let n = hist.len();
    let closes: Vec<f64> = hist.iter().map(|b| b.close).collect();
    let mut out = Vec::with_capacity(n);
    for i in 0..n {
        if i >= BB_BASE_N - 1 {
            let win = &closes[i - (BB_BASE_N - 1)..=i];
            out.push((mean(win), std_pop(win)));
        } else {
            out.push((0.0, 0.0));
        }
    }
    out
}

/// Frozen band stack + ATR14 at the setup anchor (Python `_anchor_refs`):
/// None when the anchor's 20-bar context is not fully inside the window.
struct Refs {
    mid_ref: f64,
    sd_ref: f64,
    atr_ref: f64,
    upper_1sd_ref: f64,
    upper_2sd_ref: f64,
    upper_3sd_ref: f64,
    lower_1sd_ref: f64,
    lower_2sd_ref: f64,
    lower_3sd_ref: f64,
}

fn anchor_refs(hist: &[HistBar], anchor_event_id: &str) -> Option<Refs> {
    let pos = hist.iter().position(|b| b.event_id == *anchor_event_id)?;
    if pos < BB_BASE_N - 1 || pos < 13 {
        return None;
    }
    let closes: Vec<f64> = hist.iter().map(|b| b.close).collect();
    let win = &closes[pos - (BB_BASE_N - 1)..=pos];
    let mid = mean(win);
    let sd = std_pop(win);
    let mut trs = Vec::with_capacity(14);
    for k in pos - 13..=pos {
        trs.push(hist[k].high - hist[k].low);
    }
    let atr = mean(&trs);
    Some(Refs {
        mid_ref: mid,
        sd_ref: sd,
        atr_ref: atr,
        upper_1sd_ref: mid + sd,
        upper_2sd_ref: mid + 2.0 * sd,
        upper_3sd_ref: mid + 3.0 * sd,
        lower_1sd_ref: mid - sd,
        lower_2sd_ref: mid - 2.0 * sd,
        lower_3sd_ref: mid - 3.0 * sd,
    })
}

/// Python `_direction`: (direction, ref_key) from the state's bb/ema features.
/// Fade-zone boundaries are GTE/LTE asymmetric, NOT symmetric: a-LONG is
/// `mid-3sd < close <= mid-2sd`, a-SHORT is `mid+2sd <= close < mid+3sd`
/// (the 3-SD boundary is open in both — a close beyond 3-SD is a trend).
fn direction(
    version: &str,
    variant: &str,
    close: f64,
    bb_mid: f64,
    bb_upper: f64,
    bb_lower: f64,
    ema_fast: f64,
    ema_slow: f64,
) -> (Option<&'static str>, &'static str) {
    if variant == "a" {
        if version == "v1" {
            let mid = bb_mid;
            let upper_3sd = mid + 1.5 * (bb_upper - mid);
            let lower_3sd = mid - 1.5 * (mid - bb_lower);
            if bb_upper <= close && close < upper_3sd {
                return (Some("SHORT"), "upper_2sd_ref");
            }
            if lower_3sd < close && close <= bb_lower {
                return (Some("LONG"), "lower_2sd_ref");
            }
            return (None, "");
        }
        // Tag & Re-entry: close has closed back inside the 2SD band after touching extreme
        if close < bb_upper && close > bb_mid && (bb_upper - close) < (bb_upper - bb_mid) * 0.5 {
            return (Some("SHORT"), "upper_2sd_ref");
        }
        if close > bb_lower && close < bb_mid && (close - bb_lower) < (bb_mid - bb_lower) * 0.5 {
            return (Some("LONG"), "lower_2sd_ref");
        }
        return (None, "");
    }
    if close > bb_mid && ema_fast > ema_slow {
        return (Some("LONG"), "mid_ref");
    }
    if close < bb_mid && ema_fast < ema_slow {
        return (Some("SHORT"), "mid_ref");
    }
    (None, "")
}

fn evaluate_variant(fm: &FeatMap, expert_id: &str, version: &str, variant: &str) -> ExpertEval {
    let sym = fm.symbol;
    // `_need`: close, bb_mid, bb_upper, bb_lower, bb_pct_b, ema_fast, ema_slow,
    // history — any missing required feature is NO_HABITAT (Python `_need`).
    let close = match fm.value("close") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let bb_mid = match fm.value("bb_mid") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let bb_upper = match fm.value("bb_upper") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let bb_lower = match fm.value("bb_lower") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let _bb_pct_b = match fm.value("bb_pct_b") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let ema_fast = match fm.value("ema_fast") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let ema_slow = match fm.value("ema_slow") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    if fm.history.is_empty() {
        return no_habitat(expert_id, version, fm.as_of);
    }
    let (dir, ref_key) = direction(
        version, variant, close, bb_mid, bb_upper, bb_lower, ema_fast, ema_slow,
    );
    let direction = match dir {
        Some(d) => d,
        None => return no_setup(expert_id, version, fm.as_of),
    };
    // The D-026 anchor uses the per-bar recomputed band stack (Python
    // `self._bb` from `_bb_series`), NOT the state bb_* features.
    let bb = bb_series(&fm.history);
    let pred: Box<dyn Fn(usize, &HistBar) -> bool> = match (variant, direction) {
        ("a", "LONG") => Box::new(move |i, b| {
            if i < BB_BASE_N - 1 {
                return false;
            }
            let (mid, sd) = bb[i];
            sd > 0.0 && mid - 3.0 * sd < b.close && b.close <= mid - 2.0 * sd
        }),
        ("a", "SHORT") => Box::new(move |i, b| {
            if i < BB_BASE_N - 1 {
                return false;
            }
            let (mid, sd) = bb[i];
            sd > 0.0 && mid + 2.0 * sd <= b.close && b.close < mid + 3.0 * sd
        }),
        ("b", "LONG") => Box::new(move |i, b| {
            if i < BB_BASE_N - 1 {
                return false;
            }
            b.close > bb[i].0 && b.ema_fast > b.ema_slow
        }),
        ("b", "SHORT") => Box::new(move |i, b| {
            if i < BB_BASE_N - 1 {
                return false;
            }
            b.close < bb[i].0 && b.ema_fast < b.ema_slow
        }),
        _ => unreachable!(),
    };
    let anchor = find_setup_anchor(&fm.history, &*pred);
    let refs = match anchor_refs(&fm.history, &anchor) {
        Some(r) => r,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    if refs.sd_ref <= 0.0 || refs.atr_ref <= 0.0 {
        return no_habitat(expert_id, version, fm.as_of);
    }
    // Python `_geometry`: common keys first, then the variant-specific legs.
    let mut geometry = geom(vec![
        ("entry", serde_json::json!("NEXT_BAR_CLOSE")),
        ("expiry_bars", serde_json::json!(EXPIRY_BARS)),
        ("atr_ref", serde_json::json!(refs.atr_ref)),
        ("variant", serde_json::json!(variant)),
    ]);
    if variant == "b" {
        // Setup 3: entry proxy at the 2-SD band, stop under the SMA (two
        // sigma), profit exit at the 1-SD band (one sigma).
        let stop_r = (2.0 * refs.sd_ref / refs.atr_ref).clamp(0.8, 2.0);
        let target_r = (refs.sd_ref / refs.atr_ref).clamp(0.8, 2.0);
        geometry.insert(
            "stop_r".to_string(),
            serde_json::json!(stop_r),
        );
        geometry.insert(
            "target_r".to_string(),
            serde_json::json!(target_r),
        );
        geometry.insert("mid_ref".to_string(), serde_json::json!(refs.mid_ref));
        if direction == "LONG" {
            geometry.insert(
                "upper_1sd_ref".to_string(),
                serde_json::json!(refs.upper_1sd_ref),
            );
            geometry.insert(
                "upper_2sd_ref".to_string(),
                serde_json::json!(refs.upper_2sd_ref),
            );
        } else {
            geometry.insert(
                "lower_1sd_ref".to_string(),
                serde_json::json!(refs.lower_1sd_ref),
            );
            geometry.insert(
                "lower_2sd_ref".to_string(),
                serde_json::json!(refs.lower_2sd_ref),
            );
        }
    } else {
        // Setup 2: fade the 2-SD band; the 3-SD stop and the 1-SD target are
        // each one band-sigma away.
        let r = (refs.sd_ref / refs.atr_ref).clamp(0.8, 2.0);
        geometry.insert("stop_r".to_string(), serde_json::json!(r));
        geometry.insert("target_r".to_string(), serde_json::json!(r));
        if direction == "SHORT" {
            geometry.insert(
                "upper_1sd_ref".to_string(),
                serde_json::json!(refs.upper_1sd_ref),
            );
            geometry.insert(
                "upper_2sd_ref".to_string(),
                serde_json::json!(refs.upper_2sd_ref),
            );
            geometry.insert(
                "upper_3sd_ref".to_string(),
                serde_json::json!(refs.upper_3sd_ref),
            );
        } else {
            geometry.insert(
                "lower_1sd_ref".to_string(),
                serde_json::json!(refs.lower_1sd_ref),
            );
            geometry.insert(
                "lower_2sd_ref".to_string(),
                serde_json::json!(refs.lower_2sd_ref),
            );
            geometry.insert(
                "lower_3sd_ref".to_string(),
                serde_json::json!(refs.lower_3sd_ref),
            );
        }
    }
    let ref_val = geometry.get(ref_key).and_then(|v| v.as_f64()).unwrap();
    let fingerprint = format!("{sym}:{:.6}:{:.6}", close, ref_val);
    let draft = Draft {
        direction: direction.to_string(),
        birth_time: fm.as_of,
        risk_geometry: geometry,
    };
    candidate(expert_id, version, fm.as_of, draft, anchor, fingerprint)
}

pub fn bollinger_reversion(fm: &FeatMap, expert_id: &str, version: &str) -> ExpertEval {
    // Default variant instance mirrors the Python oracle's default constructor
    // (`BollingerReversionExpert()` -> variant 'a', Setup 2).
    evaluate_variant(fm, expert_id, version, "a")
}
